File: apps/backend/src/common/sse.ts
export interface SSEEvent {
  data: string;
  event?: string;
  id?: string;
  retry?: number;
}

/**
 * Parse complete SSE events from an arbitrary network buffer.
 * The returned remainder must be retained until the next network read.
 */
export function parseSSEBuffer(buffer: string): [SSEEvent[], string] {
  const events: SSEEvent[] = [];

  while (true) {
    const separators: Array<[string, number]> = [
      ['\n\n', 2],
      ['\r\n\r\n', 4],
      ['\r\n\n', 3],
    ];
    const candidates: Array<[number, number]> = separators
      .map(([separator, length]) => [buffer.indexOf(separator), length] as [number, number])
      .filter(([index]) => index >= 0)
      .sort(([left], [right]) => left - right);

    if (candidates.length === 0) break;
    const [separatorIndex, separatorLength] = candidates[0];
    const rawEvent = buffer.slice(0, separatorIndex);
    buffer = buffer.slice(separatorIndex + separatorLength);

    const dataLines: string[] = [];
    let event: string | undefined;
    let id: string | undefined;
    let retry: number | undefined;

    for (const rawLine of rawEvent.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n')) {
      if (!rawLine || rawLine.startsWith(':')) continue;

      const separator = rawLine.indexOf(':');
      const field = separator >= 0 ? rawLine.slice(0, separator) : rawLine;
      let value = separator >= 0 ? rawLine.slice(separator + 1) : '';
      if (value.startsWith(' ')) value = value.slice(1);

      if (field === 'data') dataLines.push(value);
      else if (field === 'event') event = value;
      else if (field === 'id' && !value.includes('\u0000')) id = value;
      else if (field === 'retry' && /^\d+$/.test(value)) retry = Number(value);
    }

    if (dataLines.length > 0) events.push({ data: dataLines.join('\n'), event, id, retry });
  }

  return [events, buffer];
}

export function formatSSEEvent(event: SSEEvent): string {
  const lines: string[] = [];
  if (event.event) lines.push(`event: ${event.event}`);
  if (event.id) lines.push(`id: ${event.id}`);
  if (typeof event.retry === 'number') lines.push(`retry: ${event.retry}`);
  for (const line of event.data.split('\n')) lines.push(`data: ${line}`);
  return `${lines.join('\n')}\n\n`;
}

File: apps/ai-service/app/api/sse.py
"""Utilities for robust Server-Sent Events parsing and formatting."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class SSEEvent:
    """A parsed SSE event."""

    data: str
    event: str | None = None
    event_id: str | None = None


def format_sse(data: Any, *, event: str | None = None, event_id: str | None = None) -> str:
    """Format one complete SSE event, always terminated by a blank line."""
    payload = data if isinstance(data, str) else json.dumps(data, ensure_ascii=False)
    lines = []
    if event:
        lines.append(f"event: {event}")
    if event_id:
        lines.append(f"id: {event_id}")
    for line in payload.splitlines() or [""]:
        lines.append(f"data: {line}")
    return "\n".join(lines) + "\n\n"


def parse_sse_buffer(buffer: str) -> tuple[list[SSEEvent], str]:
    """Parse complete SSE events and return (events, incomplete_buffer).

    Network reads are arbitrary byte chunks, so callers must retain the trailing
    incomplete event rather than trying to JSON-decode each network chunk.
    """
    events: list[SSEEvent] = []
    while True:
        separator = buffer.find("\n\n")
        separator_len = 2
        if separator < 0:
            separator = buffer.find("\r\n\r\n")
            separator_len = 4
        if separator < 0:
            break

        raw_event = buffer[:separator]
        buffer = buffer[separator + separator_len :]
        data_lines: list[str] = []
        event_name: str | None = None
        event_id: str | None = None

        for raw_line in raw_event.replace("\r\n", "\n").split("\n"):
            if not raw_line or raw_line.startswith(":"):
                continue
            field, _, value = raw_line.partition(":")
            if value.startswith(" "):
                value = value[1:]
            if field == "data":
                data_lines.append(value)
            elif field == "event":
                event_name = value
            elif field == "id":
                event_id = value

        if data_lines:
            events.append(SSEEvent("\n".join(data_lines), event_name, event_id))

    return events, buffer
